grammar.js

```diff
@@ -1,311 +1,525 @@
+    tuple: ($) => seq("{", $.expr, repeat(seq(",", $.expr)), "}"),
+
+    expr: ($) =>
+      choice(
+        prec.left(1, seq(optional($.expr), $.binop, $.expr)),
+        prec(1, $.call),
+        prec(1, $.slice),
+        prec(1, $.tuple),
+        prec(1, $.identifier_preproc),
+        $.number,
+        $.bool,
+        $.lval,
+      ),
+
+    binop: (_) =>
+      choice(
+        "==",
+        "!=",
+        ">=",
+        "=",
+        "+",
+        "-",
+        "*",
+        "/",
+        "%",
+        "|",
+        "||",
+        "&",
+        "&&",
+        "&&&",
+        "<<",
+        ">>",
+        "!",
+      ),
+
+    conditional: ($) => seq($._if, optional($._elseif), optional($._else)),
+    conditional_binop: (_) => choice("&&", "|", "!"),
+
+    _if: ($) => seq("if", "(", repeat($.expr), ")", "{", repeat($.stmt), "}"),
+    _elseif: ($) =>
+      seq("else", "if", "(", repeat($.expr), ")", "{", repeat($.stmt), "}"),
+    _else: ($) => seq("else", "{", repeat($.stmt), "}"),
+
+    control_var: ($) =>
+      choice(
+        seq(
+          $.type_identifier,
+          optional(seq("<", $.type_argument_list, ">")),
+          "(",
+          repeat(seq($.expr, optional(","))),
+          ")",
+          $.identifier,
+          ";",
+        ),
+        seq(choice($._type, $.type_identifier), $.identifier, ";"),
+      ),
+
+    lval: ($) => seq($.identifier, repeat(seq(".", $.identifier))),
+
+    fval: ($) =>
       choice(
         seq(
           $.identifier,
-          repeat(seq('.', $.identifier)),
-          seq('.', $.method_identifier)
+          repeat(seq(".", $.identifier)),
+          seq(".", $.method_identifier),
         ),
         $.method_identifier,
       ),
-    ),
 
-    method: $ => seq(
-      choice($._type, $.type_identifier),
-      $.method_identifier,
-      optional(seq('<', $.type_identifier, '>')),
-      '(',
+    method: ($) =>
+      seq(
+        choice($._type, $.type_identifier),
+        $.method_identifier,
+        optional(seq("<", $.type_identifier, ">")),
+        "(",
         repeat($.parameter),
-      ')',
-      ';'
-    ),
+        ")",
+        ";",
+      ),
 
-    parameter: $ => seq(
-      optional($.direction),
-      choice($._type, $.type_identifier),
-      $.identifier,
-      optional(',')
-    ),
+    parameter: ($) =>
+      choice(
+        seq($.direction, $.identifier, optional(",")),
+        seq(
+          $.direction,
+          optional("("),
+          choice($._type, $.type_identifier),
+          optional(")"),
+          $.identifier,
+          optional(","),
+        ),
+        seq(
+          optional("("),
+          choice($._type, $.type_identifier),
+          optional(")"),
+          $.identifier,
+          optional(","),
+        ),
+      ),
 
-    direction: $ => choice('in', 'out', 'inout'),
+    direction: (_) => choice("in", "out", "inout", "packet_in", "packet_out"),
```

Comment:
`packet_in` and `packet_out` are not directions, they're [extern types](https://p4.org/p4-spec/docs/p4-16-working-draft.html#sec-packet-data-extraction) so they should be categorized as types.